gem/optimise.py

```diff
@@ -187,6 +187,12 @@ def child(expression):
     if types <= {Literal, Zero, Failure}:
         return partial_indexed(ListTensor(expressions), (index,))
 
+    if types <= {ComponentTensor, Zero}:
+        shape, = set(e.shape for e in expressions)
```

Comment:
This is like `expressions[0].shape` but safe against "bad" input? (`select_expressions` asserts that all shapes are the same, but `_select_expressions (called recursively) does not?)